optuna/visualization/_regret_bound.py

```diff
@@ -0,0 +1,176 @@
+from __future__ import annotations
+
+from typing import NamedTuple
+
+import tqdm
+
+import optuna
+from optuna._experimental import experimental_func
+from optuna.logging import get_logger
+from optuna.study.study import Study
+from optuna.terminator.erroreval import BaseErrorEvaluator
+from optuna.terminator.erroreval import CrossValidationErrorEvaluator
+from optuna.terminator.improvement.evaluator import BaseImprovementEvaluator
+from optuna.terminator.improvement.evaluator import DEFAULT_MIN_N_TRIALS
+from optuna.terminator.improvement.evaluator import RegretBoundEvaluator
+from optuna.visualization._plotly_imports import _imports
+
+
+if _imports.is_successful():
+    from optuna.visualization._plotly_imports import go
+
+_logger = get_logger(__name__)
+
+
+PADDING_RATIO = 0.05
+OPACITY = 0.25
+
+
+class _RegretBoundInfo(NamedTuple):
+    trial_numbers: list[int]
+    regret_bounds: list[float]
+    errors: list[float] | None
+
+
+@experimental_func("3.2.0")
+def plot_regret_bound(
+    study: Study,
+    plot_error: bool = False,
+    improvement_evaluator: BaseImprovementEvaluator | None = None,
+    error_evaluator: BaseErrorEvaluator | None = None,
+    min_n_trials: int = DEFAULT_MIN_N_TRIALS,
+) -> "go.Figure":
+    _imports.check()
+
+    info = _get_regret_bound_info(study, plot_error, improvement_evaluator, error_evaluator)
+    return _get_regret_bound_plot(info, min_n_trials)
+
+
+def _get_regret_bound_info(
+    study: Study,
+    get_error: bool = False,
+    improvement_evaluator: BaseImprovementEvaluator | None = None,
+    error_evaluator: BaseErrorEvaluator | None = None,
+) -> _RegretBoundInfo:
+    if improvement_evaluator is None:
+        improvement_evaluator = RegretBoundEvaluator()
+    if error_evaluator is None:
+        error_evaluator = CrossValidationErrorEvaluator()
+
+    trial_numbers = []
+    regret_bounds = []
+    errors = []
+
+    for i, trial in enumerate(tqdm.tqdm(study.trials)):
+        trial_numbers.append(trial.number)
+        trials = study.trials[: i + 1]
+        trials = [t for t in trials if t.state == optuna.trial.TrialState.COMPLETE]
+
+        regret_bound = improvement_evaluator.evaluate(
+            trials=trials, study_direction=study.direction
+        )
+        regret_bounds.append(regret_bound)
+
+        if get_error:
+            error = error_evaluator.evaluate(trials=trials, study_direction=study.direction)
```

Comment:
[Question] Currently, the error handling seems to be delegated to the `ErrorEvaluator` when `plot_error` is `True` and `error_evaluator` is `None`. Do you think it is better to handle errors in this function?

Reply:
In my opinion, current implementation is reasonable because `ErrorEvaluator` already handle such error appropriately and to handle errors for various `ErrorEvaluator` in this function is not so clear approach. What do you think?

Reply:
Being nitpicky, the `BaseErrorEvaluator` class seems not to be responsible for the error handling when the error scores are not reported. The child class, `CrossValidationErrorEvaluator`, coincidentally handles it in the current implementation.

However, I think it's OK to ignore those details so far. We can revisit it when we implement another subclass of `BaseErrorEvaluator`.